File: DSAlgorithms/Leetcode/randoms/206-reverseLinkedList.cpp
struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

ListNode* reverseList(ListNode* head)
{
    ListNode* prev = nullptr;
    ListNode* current = head;
    ListNode* next = nullptr;
    while(current != nullptr)
    {
        next = current->next;
        current->next = prev;
        prev = current; 
        current = next;
    }
    return prev;
}

ListNode* reverseListRecursive(ListNode* head)
{
    if(head == nullptr)
        return nullptr;
    ListNode* newHead = head;
    if(head->next)
    {
        newHead = reverseListRecursive(head->next);
        head->next->next = head;
    }
    head->next = nullptr;
    return newHead;
}

File: dsa_algo/leetcode/ts/src/old/206-reverseLinkedlist.ts
const reverseList = (head: ListNode | null): ListNode | null => {
    let prev: ListNode | null = null;
    let next: ListNode | null = null;
    let current: ListNode | null = head;
    while(current !== null)
    {
        next = current.next;
        current = prev;
        prev = current;
    }
    return prev;
}
